### PR TITLE
[ENH] Add new `FunctionParamFitter` parameter estimator

### LINKED ISSUE #4818: [ENH] conditional forecaster multiplexing by instance property, e.g., slow or fast moving goods
**Is your feature request related to a problem? Please describe.**
Here is the logic that I would like to use with sktime:
* I have some panel data, where the data is grouped into instances by 'store' and 'product', and the target value is 'Sales'
* I have a special rule that does a different type of forecast based on the product and past sales (it predicts all zeros based on a heuristic that predicts if a store no longer stocks a product)

Now, I have my custom forecaster working well, and its `fit` method is called with `X` for each instance, but there's no way for me to tell which instance I'm currently looking at (that is, which store/product I'm looking at) so I can't apply any product-specific logic.

**Describe the solution you'd like**
I've looked at the code and can't see how it could be done easily. In this code, `group_name` would need to be passed through to the estimator method, but of course that would error if it's not expected. You could `inspect` the func to see if it's expected.
https://github.com/sktime/sktime/blob/f94a4bf4ed6e7ec1520496ec00ae987c57cc9703/sktime/datatypes/_vectorize.py#L572-L584

**Describe alternatives you've considered**
A few workarounds that work just fine:
* Add a column to X `IsProductSomething` so the information is in X
* Put all the logic for whether or not to do a particular type of forecast outside the sktime pipeline, so have a boolean column for `JustPredictZero` and read that in the forecaster.

These both spread out the logic of conditional forecasting which isn't ideal.

I could also put the Product column in X (in addition to being in the index), but it's a string, so then I need to mess around with categories/integer representations, or drop the column before passing to `fit`, and it needs a different name (else it clashes with the index) so that approach is a bit of a mess.

**Additional context**
This is probably quite a niche problem, I'd be perfectly happy for you to close this. But if at some point the opportunity arises to make this happen that would be neat.

### PR DESCRIPTION
**Pull request by tpvasconcelos**
#### Reference Issues/PRs

Fixes #4818

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

Note that this implementation was heavily inspired by the existing [FunctionTransformer](https://github.com/sktime/sktime/blob/9daa6d718569187b541161b0065e1fdf5bd7afe6/sktime/transformations/series/func_transform.py) implementation.

#### Did you add any tests for the change?

Doctests only

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/param_est/base.py [hunk 1]
```diff
@@ -396,8 +396,8 @@ def _get_fitted_params(self):
...
-        fitted_params = [attr for attr in dir(self) if attr.endswith("_")]
-        fitted_params = [x for x in fitted_params if not x.startswith("_")]
-        fitted_param_dict = {p[:-1]: getattr(self, p) for p in fitted_params}
-
-        return fitted_param_dict
+        return {
+            attr[:-1]: getattr(self, attr)
+            for attr in dir(self)
+            if attr.endswith("_") and not attr.startswith("_")
+        }
```
Comment (tpvasconcelos) [inline sktime/param_est/base.py:403]: Note to reviewer: This is just a very small refactor to make this piece of code easier to follow and a bit more efficient

Comment (fkiraly) [reply sktime/param_est/base.py:403]: hm, I forgot this was there. Good spot.

Because there is some separate discussion to have here - if this does not change logic, would you be so kind and move this refactor to a separate PR?

To give some explanation:

* in my internal architectural model, this actually lives in `BaseEstimator`. Quick inspection indeed indicates that the parent base class `BaseEstimator` seems to have an alternative, perhaps better version of the code.
* this would suggest instead the action to delete this code, and potentially refactor the `BaseEstimator` code instead
* the explanation for the duplication might be, we first introduced `get_fitted_params` for parameter fitters, but then realized it makes sense in all estimators.
* mid-term, the plan was to use `scikit-base` `BaseEstimator` to get `get_fitted_params  functionality, where the code *also* should live, in the `BaseEstimator` there. So, a mid-term action would be to remove duplicated code in `sktime` `BaseEstimator` and check `scikit-base` whether it should get an update.

So, to summarize, may I suggest to separate this refactor out in a PR, and make a note in an issue to possibly refactor this locus across three locations - `scikit-base` `BaseEstimator`, `sktime` `BaseEstimator`, and `BaseParamFitter`?

File: sktime/param_est/compose.py [hunk 2]
```diff
@@ -301,3 +302,120 @@ def get_test_params(cls, parameter_set="default"):
...
+
+class FunctionParamFitter(BaseParamFitter):
+    r"""Constructs a parameter fitter from an arbitrary callable.
+
+    A FunctionParamFitter forwards its X argument to a user-defined
+    function (or callable object) and sets the result of this function
+    to the ``param`` attribute. This can be useful for stateless
+    estimators such as simple conditional parameter selectors.
+
+    Note: If a lambda function is used as the ``func``, then the
```
Comment (fkiraly) [inline sktime/param_est/compose.py:315]: yes, indeed, it is sth I wanted to discuss.

For now, the design makes sense, I just pointed out that you can define a custom `save` function and `load_from_` functions.

We have a couple other cases where `lambda`-s can be parameters, and we have not found a good way to make these serializable. In case you have such good ideas, we are all ears. FYI @achieveordie

File: sktime/param_est/compose.py [hunk 3]
```diff
@@ -301,3 +302,120 @@ def get_test_params(cls, parameter_set="default"):
...
+    resulting estimator will not be pickleable.
+
+    Parameters
+    ----------
+    param : str
+        The name of the parameter to set.
+    func : callable (X: X_type, **kwargs) -> Any
+        The callable to use for the parameter estimation. This will be
+        passed the same arguments as estimator, with args and kwargs
+        forwarded.
```
Comment (fkiraly) [inline sktime/param_est/compose.py:325]: some params are missing here, but all make sense

**Review by fkiraly — CHANGES_REQUESTED** [review]
The code is great and I would approve it once docs are complete. Example in the docstring is also very nice and clear. Let's also separate out the base class refactor, this is a very good spot but separate topic and likely to lead to longer discussion

Docs:

* should be added to API reference, `docs/source/api_reference`
* docstring is incomplete, params are missing

Suggestions for improvement, non-blocking:

* in the example, you could also use `get_fitted_params` to look at the `selected_forecaster`, and carry out the combination with the `MultiplexForecaster` to have a form of custom tuning!
* let's not have too many estimators in one file, especially the `compose` modules have gotten bloated in the past. I would suggest, turn it into a folder with private files and export at the `compose` module level, similar to `forecasting.compose`.

File: sktime/param_est/compose.py [hunk 4]
```diff
@@ -301,3 +302,120 @@ def get_test_params(cls, parameter_set="default"):
...
+    >>> param_est = FunctionParamFitter(
+    ...     param="selected_forecaster",
+    ...     func=(
+    ...         lambda X, threshold: "naive-seasonal"
+    ...         if len(X) >= threshold
+    ...         else "naive-last"
+    ...     ),
+    ...     kw_args={"threshold": 7},
+    ... )
+    >>> param_est.fit(np.asarray([1, 2, 3, 4]))
```
Comment (fkiraly) [inline sktime/param_est/compose.py:352]: failure is due to how doctest work, it compares the printouts against what is below the line. Here, it expects `FucntionParamFitter(...)`, as the return of `fit` is self. You can use the ellipsis arg `...` to catch the param printout which may format differently by operating system.

File: sktime/transformations/series/func_transform.py [hunk 5]
```diff
@@ -18,12 +18,12 @@ class FunctionTransformer(BaseTransformer):
     r"""Constructs a transformer from an arbitrary callable.
 
     A FunctionTransformer forwards its y (and optionally X) arguments to a
-    user-defined function or function object and returns the result of this
+    user-defined function (or callable object) and returns the result of this
```
Comment (fkiraly) [inline sktime/transformations/series/func_transform.py:21]: small enough to have it hide in another PR, though my general recommendation is to make separate PR for sth like this, it is one that would be instant-accepted and merged. The danger is that insta-merge ride-along changes get stalled with the host PR and clutter its review.
(you also hurt your honest contribution statistics, if that were something you care about)

File: sktime/param_est/base.py (hunk 1 above)
Comment (tpvasconcelos) [reply sktime/param_est/base.py]: Reverted these changes and moved them to #5633 (referenced this comment there)

File: sktime/transformations/series/func_transform.py (hunk 5 above)
Comment (tpvasconcelos) [reply sktime/transformations/series/func_transform.py:21]: Good point! Moved these changes to #5634

File: sktime/param_est/compose/_func_fitter.py
Comment (tpvasconcelos) [inline sktime/param_est/compose/_func_fitter.py:1]: Naming is hard... do you have a preference for this module's name? 😄

File: sktime/param_est/compose/_func_fitter.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        Parameters
+        ----------
+        X : guaranteed to be of a type in self.get_tag("X_inner_mtype")
+            Time series to which to fit the estimator.
+
+        Returns
+        -------
+        self : reference to self
+        """
+        param = self.param.rstrip("_") + "_"
```
Comment (tpvasconcelos) [inline sktime/param_est/compose/_func_fitter.py:104]: @fkiraly Just double checking in case you missed this one. This decision is not explicitly documented or explained anywhere to the user. Still, this line implicitly makes `'some_param'`, `'some_param_'`, and even `'some_param___'` to be treated as the same parameter (i.e. they are all automatically converted to `'some_param_'` here). There are a few options to consider here:
1. Allow the user to pass any variation of `'some_param'`, `'some_param_'`, `'some_param___'`, etc. and automatically strip and convert to `'some_param_'` (**the current implementation in the PR**)
2. Only allow parameter name without a trailing `_`. i.e., raise a value error if `param.endswith('_')`
3. Require the user to pass the parameter name with the trailing `_`
4. Something else...?

I think that the first two options make the most sense but I'm more inclined towards the first one as it creates the least friction for end users (i.e., _it just works_)

File: sktime/param_est/compose.py (hunk 2 above)
Comment (tpvasconcelos) [reply sktime/param_est/compose.py]: Without having thought too much about this, I am initially inclined to say that this does not sound like something that sktime should try to solve. 

I believe that there are mature tools out there that _could_ be used to serialise lambdas (for instance, https://github.com/uqfoundation/dill). After a quick Google search, I saw that even using the builtin `marcal` package, you can do a simple round trip with:
```pycon
>>> my_lambda = lambda: "Hello"
>>> my_lambda_roundtrip = types.FunctionType(marshal.loads(marshal.dumps(my_lambda.__code__)), globals())
>>> my_lambda_roundtrip()
'Hello'
```

However, I could not find a reliable (**and simple!**) way to **make** a lambda pickleable.

Even if there was it sounds strange for sktime to do this behind the scenes even when it might not be necessary to do so, adding unnecessary overhead to all constructors that take callables as arguments.

Resolving this comment for now as it is not relevant to this PR but interesting in a follow up on this 👍

File: sktime/param_est/compose.py (hunk 4 above)
Comment (tpvasconcelos) [reply sktime/param_est/compose.py]: Good catch! Fixed ✅

**Comment by tpvasconcelos** [comment]
@fkiraly Thanks for all the feedback! I belive all points have been addressed. I will open this PR for a full review now :)

**Comment by tpvasconcelos** [comment]
For reference, see the rendered docs here: https://sktime--5630.org.readthedocs.build/en/5630/api_reference/auto_generated/sktime.param_est.compose.FunctionParamFitter.html

File: sktime/param_est/compose/_func_fitter.py
Comment (fkiraly) [reply sktime/param_est/compose/_func_fitter.py:1]: no, it is anyway private

File: sktime/param_est/compose/_func_fitter.py (hunk 6 above)
Comment (fkiraly) [reply sktime/param_est/compose/_func_fitter.py:135]: hm, I think it should be consistent with the name in `get_fitted_params`, so the current implementation makes sense.

**Review by fkiraly — APPROVED** [review]
Yes, changes have been addressed. Thanks!

**Comment by fkiraly** [comment]
One non-blocking comment, I am slightly wondering about the case of multiple fitted parameters, but none of the ways in which one could generalize makes me happy, so I would suggest to leave the signature as is.

**Comment by tpvasconcelos** [comment]
> One non-blocking comment, I am slightly wondering about the case of multiple fitted parameters, but none of the ways in which one could generalize makes me happy, so I would suggest to leave the signature as is.

hmm what about something like this:

```python
class MultiParamFitter(BaseParamFitter):
    def __init__(self, fitters: dict[str, Callable]):
        self.fitters = fitters

    def _fit(self, X):
        for param, fitter in self.fitters.items():
            param = param.rstrip("_") + "_"
            setattr(self, param, self.func(X))
        return self
```

**Comment by fkiraly** [comment]
yes, thought of that - do not like it since the user has to pass an iterable, which is confusing in the single fitter case.

Other option is returning `dict`, which is confusing in the single parameter case...
